lib/cuda/texture.jl

```diff
@@ -0,0 +1,402 @@
+                       texarr.handle, # 1 x UInt64
+                       ntuple(_->Int64(0), 15), # 15 x UInt64
+                       UInt32(0)))
+    return resDesc_ref
+end
+
+# allow passing mismatching refs where we're expecting a CUDA_RESOURCE_DESC
+Base.unsafe_convert(::Type{Ptr{CUDA_RESOURCE_DESC}}, ref::Base.RefValue{T}) where {T} =
+    convert(Ptr{CUDA_RESOURCE_DESC}, Base.unsafe_convert(Ptr{T}, ref))
+
+function CUDA_RESOURCE_DESC(arr::CuArray{T,N}, channels, format) where {T,N}
+    # TODO: take care of allowed pitches
+    1 <= N <= 2 || throw(ArgumentError("Only 1 or 2D CuArray objects can be wrapped in a texture"))
+
+    # FIXME: manual construction due to invalid padding (JuliaInterop/Clang.jl#238)
+    resDesc_ref = Ref(((N == 1 ? CU_RESOURCE_TYPE_LINEAR : CU_RESOURCE_TYPE_PITCH2D), # resType::CUresourcetype
+                       pointer(arr), # 1 x UInt64 (CUdeviceptr)
+                       format, # 1/2 x UInt64 (CUarray_format)
+                       UInt32(channels), # 1/2 x UInt64
+                       (N == 2 ? size(arr, 1) : size(arr, 1) * sizeof(T)), # 1 x UInt64 nx
+                       (N == 2 ? size(arr, 2) : 0), # 1 x UInt64 ny
+                       (N == 2 ? size(arr, 1) * sizeof(T) : 0), # 1 x UInt64 pitch
+                       ntuple(_->Int64(0), 11), # 11 x UInt64
+                       UInt32(0)))
+    return resDesc_ref
+end
+
+function unsafe_destroy!(t::CuTexture)
+    if isvalid(t.ctx)
+        cuTexObjectDestroy(t)
+    end
+end
+
+Base.convert(::Type{CUtexObject}, t::CuTexture) = t.handle
+
+
+## array interface
+
+Base.size(tm::CuTexture) = size(tm.parent)
+
+Base.eltype(tm::CuTexture{T,N,1}) where {T,N} = T
+Base.eltype(tm::CuTexture{T,N,C}) where {T,N,C} = NTuple{C,T}
+
+
+## interop with other arrays
+
+CuTexture{T,N,C}(x::Union{CuArray{T,N},CuTextureArray{T,N,C}}; kwargs...) where {T,N,C} =
+    CuTexture{T,N,C,typeof(x)}(x; kwargs...)
+
+"""
+    CuTexture(x::CuTextureArray{T,N,C})
+
+Create a `N`-dimensional texture object with each `C` channels of elements of type `T` that
+will be read from `x`.
+"""
+CuTexture(x::CuTextureArray{T,N,C}; kwargs...) where {T,N,C} =
+    CuTexture{T,N,C}(x; kwargs...)
+
+"""
+    CuTexture(x::CuArray{T,N})
+
+Create a `N`-dimensional texture object that reads from a `CuArray`. If the `T` is an
+`NTuple`, the texture element type and number of channels will be determined from it.
+Otherwise, the array element type is used as-is and the texture object will have one channel.
+
+Note that it is necessary the their memory is well aligned and strided (good pitch).
+Currently, that is not being enforced.
+"""
+CuTexture(x::CuArray{T,N}; kwargs...) where {T,N} =
+    CuTexture{T,N,1}(x; kwargs...)
+CuTexture(x::CuArray{NTuple{C,T},N}; kwargs...) where {T,N,C} =
+    CuTexture{T,N,C}(x; kwargs...)
+
+Adapt.adapt_storage(::Adaptor, t::CuTexture{T,N,C}) where {T,N,C} =
+    CuDeviceTexture{T,N,C,t.normalized_coordinates}(size(t), t.handle)
```

Comment:
(copying comment from https://github.com/JuliaGPU/CUDA.jl/pull/206)
Can this bring any type of instability?
I have not used a solution like this in the first place due to the fear of type instability, but now I guess that maybe the kernel launch works as a function barrier, is that right?

Reply:
Yes, it works like a function barrier indeed, we never get type instabilities across the host<->device barrier (unless we legitimately were dealing with an abstract type.